crypto/x509/x509_vfy.c

```diff
@@ -1754,119 +1754,70 @@ int X509_cmp_current_time(const ASN1_TIME *ctm)
-        int min_length = sizeof("YYMMDDHHMMZ") - 1;
-        int max_length = sizeof("YYMMDDHHMMSS+hhmm") - 1;
-        if (remaining < min_length || remaining > max_length)
-            return 0;
-        memcpy(p, str, 10);
-        p += 10;
-        str += 10;
-        remaining -= 10;
-    } else {
-        /* YYYYMMDDHHMM[SS[.fff]]Z or YYYYMMDDHHMM[SS[.f[f[f]]]](+-)hhmm */
-        int min_length = sizeof("YYYYMMDDHHMMZ") - 1;
-        int max_length = sizeof("YYYYMMDDHHMMSS.fff+hhmm") - 1;
-        if (remaining < min_length || remaining > max_length)
-            return 0;
-        memcpy(p, str, 12);
-        p += 12;
-        str += 12;
-        remaining -= 12;
-    }
-
-    if ((*str == 'Z') || (*str == '-') || (*str == '+')) {
-        *(p++) = '0';
-        *(p++) = '0';
-    } else {
-        /* SS (seconds) */
-        if (remaining < 2)
+    switch (ctm->type) {
+    case V_ASN1_UTCTIME:
+        if (ctm->length != utctime_length)
             return 0;
-        *(p++) = *(str++);
-        *(p++) = *(str++);
-        remaining -= 2;
         /*
-         * Skip any (up to three) fractional seconds...
-         * TODO(emilia): in RFC5280, fractional seconds are forbidden.
-         * Can we just kill them altogether?
+         * Years >= 50 are 19YY, years < 50 are 20YY.
+         * We verify all digits later; for now we assume they're in valid range.
          */
-        if (remaining && *str == '.') {
-            str++;
-            remaining--;
-            for (i = 0; i < 3 && remaining; i++, str++, remaining--) {
-                if (*str < '0' || *str > '9')
-                    break;
-            }
-        }
-
+        year = (ctm->data[0] - '0') * 10 + ctm->data[1] - '0';
+        memcpy(time1, year >= 50 ? "19" : "20", 2);
+        memcpy(time1 + 2, ctm->data, ctm->length);
+        break;
+    case V_ASN1_GENERALIZEDTIME:
+        if (ctm->length != generalizedtime_length)
+            return 0;
+        memcpy(time1, ctm->data, ctm->length);
+        break;
+    default:
+        return 0;
     }
-    *(p++) = 'Z';
-    *(p++) = '\0';
 
-    /* We now need either a terminating 'Z' or an offset. */
-    if (!remaining)
-        return 0;
-    if (*str == 'Z') {
-        if (remaining != 1)
-            return 0;
-        offset = 0;
-    } else {
-        /* (+-)HHMM */
-        if ((*str != '+') && (*str != '-'))
-            return 0;
-        /* Historical behaviour: the (+-)hhmm offset is forbidden in RFC5280. */
-        if (remaining != 5)
-            return 0;
-        if (str[1] < '0' || str[1] > '9' || str[2] < '0' || str[2] > '9' ||
-            str[3] < '0' || str[3] > '9' || str[4] < '0' || str[4] > '9')
+    /**
+     * Now everything is in GeneralizedTime. We verify that the digits
+     * are in valid range; we don't verify that the date is valid.
+     */
+    for (size_t i = 0; i < generalizedtime_length - 1; i++) {
+        if (time1[i] < '0' || time1[i] > '9')
```

Comment:
This type of code is all over the place; I think it would make sense to do that, however.